File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.py
import heapq
import sys
def input():return sys.stdin.readline().rstrip()
dx=[0,0,1,-1,-1,1,1,-1]
dy=[1,-1,0,0,1,1,-1,-1]
idx=1
while 1:
    N=int(input())
    if not N:
        break
    arr=[list(map(int,input().split())) for _ in range(N)]
    q=[]
    q.append((arr[0][0],0,0))
    dist = [[987654321 for _ in range(126)] for _ in range(126)]
    while q:
        cost,x,y=heapq.heappop(q)
        for i in range(4):
            n_x=x+dx[i]
            n_y=y+dy[i]
            if 0<=n_x<N and 0<=n_y<N:
                nc=cost + arr[n_x][n_y]
                if dist[n_x][n_y] > nc:
                    dist[n_x][n_y]=nc
                    q.append((nc,n_x,n_y))
    print("Problem %d: %d"%(idx,dist[N-1][N-1]))
    idx+=1

File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.java
import java.io.*;
import java.sql.Array;
import java.util.*;

public class Main {
    static int dx[]={0,0,1,-1};
    static int dy[]={1,-1,0,0};
    static class Info implements Comparable<Info> {
        int x,y,cost;
        public Info(int x,int y,int cost){
            this.x=x;
            this.y=y;
            this.cost=cost;
        }
        @Override
        public int compareTo(Info o)
        {
            return this.cost-o.cost;
        }

    }
    public static boolean is_inside(int x,int y,int N,int M)
    {
        return 0<= x && x<N &&0<=y &&y<M;
    }
    static int arr[][];
    static int dist[][];
    static int N,M,K,X;
    public static void main(String[] args) throws IOException {
        X=0;
        BufferedReader br= new BufferedReader(new InputStreamReader(System.in));
        while (true)
        {
            X+=1;
            N=Integer.parseInt(br.readLine());
            if (N==0)
                break;
            arr=new int[N+1][N+1];
            dist=new int[N+1][N+1];
            for (int i=0;i<N+1;i++)
                Arrays.fill(dist[i],987654321);
            for (int i=0;i<N;i++)
            {
                StringTokenizer st=new StringTokenizer(br.readLine());
                for (int j=0;j<N;j++)
                {
                    arr[i][j]=Integer.parseInt(st.nextToken());
                }
            }
            PriorityQueue<Info> q= new PriorityQueue<>();
            q.add(new Info(0,0,arr[0][0]));
            dist[0][0]=arr[0][0];
            while (!q.isEmpty())
            {
                Info info=q.poll();
                int x=info.x,y=info.y,cost=info.cost;
                if (dist[x][y]<cost)continue;
                for (int i=0;i<4;i++)
                {
                    int n_x=x+dx[i];
                    int n_y=y+dy[i];
                    if (is_inside(n_x,n_y,N,N) && dist[n_x][n_y]>cost+arr[n_x][n_y])
                    {
                        dist[n_x][n_y]=cost+arr[n_x][n_y];
                        q.add(new Info(n_x,n_y,cost+arr[n_x][n_y]));
                    }
                }
            }
            System.out.println("Problem " + X + ": " + dist[N-1][N-1]);
        }
    }
}